core/src/graph/shared_state_manager.rs

```diff
@@ -146,41 +151,33 @@ impl PublicKeyProvider for SharedStateManager {
 		Ok(())
 	}
 
-	fn export_new_keys(&self) -> Result<Vec<DsnpKeys>> {
+	fn export_new_key_updates(&self) -> Result<Vec<Update>> {
 		let mut result = vec![];
 		for (dsnp_user_id, key) in self.new_keys.inner() {
-			result.push(DsnpKeys {
-				dsnp_user_id: *dsnp_user_id,
-				keys: vec![KeyData {
-					content: Frequency::write_public_key(&key)
-						.map_err(|e| Error::msg(format!("failed to serialize key {:?}", e)))?,
-					// all new keys are assigned a new id so we can unwrap here
-					index: key.key_id.unwrap() as u16,
-				}],
-				keys_hash: self
-					.dsnp_user_to_keys
-					.get(&dsnp_user_id)
-					.expect("Key hash should exist")
-					.1,
+			let prev_hash = self
+				.dsnp_user_to_keys
+				.get(&dsnp_user_id)
+				.map_or(PageHash::default(), |(_, hash)| *hash);
+			result.push(Update::AddKey {
+				owner_dsnp_user_id: *dsnp_user_id,
+				prev_hash,
+				payload: Frequency::write_public_key(key)?,
 			});
 		}
 		Ok(result)
 	}
 
-	fn get_all_keys(&self, dsnp_user_id: DsnpUserId) -> Vec<&DsnpPublicKey> {
+	fn get_imported_keys(&self, dsnp_user_id: DsnpUserId) -> Vec<&DsnpPublicKey> {
 		let mut all_keys = vec![];
 		if let Some((v, _)) = self.dsnp_user_to_keys.get(&dsnp_user_id) {
 			all_keys.extend(&v[..]);
 		}
-		if let Some(k) = self.new_keys.get(&dsnp_user_id) {
```

Comment:
removing new keys from here since they can not being used until being publish on chain